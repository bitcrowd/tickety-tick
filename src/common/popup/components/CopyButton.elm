module CopyButton exposing (..)

import Html exposing (..)
import Html.Attributes exposing (..)
import Html.Events exposing (..)

import Messages exposing (Msg(Grab))


view : List (Attribute Msg) -> List (Html Msg) -> String -> Html Msg
view attributes children val =
  button (attributes ++ [ value val, onClick (Grab val) ]) children
