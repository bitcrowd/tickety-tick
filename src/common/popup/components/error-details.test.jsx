import React from 'react';
import { shallow } from 'enzyme';
import StackTrace from 'stacktrace-js';

import CopyErrorDetails from './error-details';
import CopyButton from './copy-button';

jest.mock('stacktrace-js', () => ({ fromError: jest.fn() }));

class MockStackFrame {
  constructor(functionName, fileName, lineNumber, columnNumber) {
    this.functionName = functionName;
    this.fileName = fileName;
    this.lineNumber = lineNumber;
    this.columnNumber = columnNumber;
  }

  getFunctionName() {
    return this.functionName;
  }

  getFileName() {
    return this.fileName;
  }

  getLineNumber() {
    return this.lineNumber;
  }

  getColumnNumber() {
    return this.columnNumber;
  }
}

describe('error-details', () => {
  function render(overrides) {
    const defaults = { errors: [new Error('WAT?')] };
    const props = { ...defaults, ...overrides };
    return shallow(<CopyErrorDetails {...props} />);
  }

  beforeEach(() => {
    StackTrace.fromError.mockResolvedValue([
      ['scan', 'src/common/adapters/github.js', 74, 8],
      ['attempt', 'src/common/search.js', 12, 26],
      ['search', 'src/common/search.js', 27, 45],
      ['search', 'src/common/search.js', 27, 25],
      ['browser.runtime.onMessage.addListener', 'src/web-extension/content.js', 8, 13],
      ['wrappedSendResponse', 'node_modules/webextension-polyfill/dist/browser-polyfill.js', 1057, null],
    ].map(args => new MockStackFrame(...args)));
  });

  afterEach(() => {
    StackTrace.fromError.mockClear();
  });

  it('renders a button to copy the error details', async () => {
    const wrapper = render({ errors: [new Error('Boom!')] });
    await new Promise(resolve => setTimeout(resolve, 0)); // wait for stacktrace-js processing
    expect(wrapper.find(CopyButton).prop('value')).toMatchSnapshot();
  });
});
