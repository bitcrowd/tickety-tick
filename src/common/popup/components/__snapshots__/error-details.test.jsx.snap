// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error-details renders a button to copy the error details 1`] = `
"Tickety-Tick revision: test-commit-hash

\`\`\`
Boom!
    at scan (src/common/adapters/github.js:74:8)
    at attempt (src/common/search.js:12:26)
    at search (src/common/search.js:27:45)
    at search (src/common/search.js:27:25)
    at browser.runtime.onMessage.addListener (src/web-extension/content.js:8:13)
    at wrappedSendResponse (node_modules/webextension-polyfill/dist/browser-polyfill.js:1057:-)
\`\`\`"
`;
